pkg/boot/universalpayload/universalpayload.go

```diff
@@ -0,0 +1,449 @@
+	}
+
+	if length != (uint64)(buf.Len()-prev) {
+		return fmt.Errorf("length mismatch when appending acpi table")
+	}
+
+	*hobLen += length
+
+	return nil
+}
+
+func appendSmbiosTableHob(buf *bytes.Buffer, hobLen *uint64) error {
+	// Construct SMBIOS Hob
+	smbiosTable, err := constructSmbiosTable()
+	if err != nil {
+		return err
+	}
+
+	smbiosTableGUIDHob, err := constructGUIDHob(UniversalPayloadSmbiosTableGUID)
+	if err != nil {
+		return err
+	}
+
+	length := uint64(unsafe.Sizeof(EfiHobGUIDType{}) + unsafe.Sizeof(UniversalPayloadSmbiosTable{}))
+	prev := buf.Len()
+
+	if err := binary.Write(buf, binary.LittleEndian, smbiosTableGUIDHob); err != nil {
+		return fmt.Errorf("failed to append smbios table guid to buffer")
+	}
+
+	if err := binary.Write(buf, binary.LittleEndian, smbiosTable); err != nil {
+		return fmt.Errorf("failed to append smbios table to buffer")
+	}
+
+	// TODO: Remove this hardcode 4
+	// This hardcode is introduced by golang structure algnment vs. write bytes.
+	var pad [4]byte
+	if err := binary.Write(buf, binary.LittleEndian, pad); err != nil {
+		return fmt.Errorf("failed to append smbios table pad to buffer")
+	}
+
+	if length != (uint64)(buf.Len()-prev) {
+		return fmt.Errorf("length mismatch when appending smbios table")
+	}
+
+	*hobLen += length
+
+	return nil
+}
+
+func appendEfiCPUHob(buf *bytes.Buffer, hobLen *uint64) error {
+	cpuHob, err := hobCreateEfiHobCPU()
+	if err != nil {
+		return err
+	}
+
+	length := uint64(unsafe.Sizeof(EfiHobCPU{}))
+	prev := buf.Len()
+	if err := binary.Write(buf, binary.LittleEndian, cpuHob); err != nil {
+		return fmt.Errorf("failed to append cpu hob to buffer")
+	}
+
+	if length != (uint64)(buf.Len()-prev) {
+		return fmt.Errorf("length mismatch when appending cpu hob")
+	}
+
+	*hobLen += length
+
+	return nil
+}
+
+func constructHobList(dst *bytes.Buffer, src *bytes.Buffer, hobLen *uint64) error {
+	handoffHob := hobCreateEfiHobHandoffInfoTable(*hobLen)
+	if err := binary.Write(dst, binary.LittleEndian, handoffHob); err != nil {
+		return fmt.Errorf("failed to append handoff hob to buffer")
+	}
+
+	if err := binary.Write(dst, binary.LittleEndian, src.Bytes()); err != nil {
+		return fmt.Errorf("failed to append hos list to buffer")
+	}
+
+	hobEndHeader := hobCreateEndHob()
+	prev := dst.Len()
+	length := uint64(unsafe.Sizeof(EfiHobGenericHeader{}))
+
+	if err := binary.Write(dst, binary.LittleEndian, hobEndHeader); err != nil {
+		return fmt.Errorf("failed to append end hob")
+	}
+
+	if length != (uint64)(dst.Len()-prev) {
+		return fmt.Errorf("length mismatch when appending end hob")
+	}
+
+	*hobLen += length
+
+	return nil
+}
+
+func loadKexecMemWithHobs(fdtLoad *FdtLoad, fdtData []byte) (kexec.Memory, error) {
+	//Step 1, Prepare memory
+	mem := kexec.Memory{}
+
+	//Step 2, Insert tianocore raw binary
+	mem.Segments.Insert(kexec.NewSegment(fdtData, kexec.Range{Start: uintptr(fdtLoad.Load), Size: uint(len(fdtData))}))
```

Comment:
is this safe? how do we know this range isn't some reserved piece of memory?

Reply:
Good question.
Actually, UPL FIT image is generated by EDK2 source code, and the expected load address of UPL FIT imaged is configured by EDK2 developer. Which means, EDK2 developer should have knowledge about memory layout of their own hardware platform, for instance, MMIO region, System RAM region etc. So I think the expected load address of UPL FIT image should be reasonable and safe to use.

Reply:
The spec says the following: https://universalpayload.github.io/spec/chapter2-payload-image-format.html#fit-schema

> This should not normally be needed, since Payloads should be loadable to any suitable address.
> 
> Where that is not possible: This is the required load address for the image. Value Type matches the machine word size. This must be provided for the ‘firmware’ image in a configuration, but is optional for others. Images without a load address can be loaded to any suitable location. They can also be left where they are in the FIT, i.e. not loaded at all.

For edk2 -- is edk2 not relocatable?

Reply:
Thanks for providing details about this load address from spec.

EDK2 has performed a relocation for univserial payload already during generating UPL FIT image here:
https://github.com/tianocore/edk2/blob/master/UefiPayloadPkg/UniversalPayloadBuild.py#L254

Back to your question about "is edk2 payload not relocatable?":
Not sure whether relocatable is supported or not. But it should support relocation.

An alternative solution is to init the kexec.Memory with system RAM regions that are read from /sys/firmware/memory. In this way, we can ensure the loaded address of payload is safe. But we still need to check whether universal payload image supports position independent feature or not. If yes, we can load it to any available memory location directly. Or, we need to perform a symbol relocation when relocating payload.

Reply:
> EDK2 has performed a relocation for univserial payload already during generating UPL FIT image here:
https://github.com/tianocore/edk2/blob/master/UefiPayloadPkg/UniversalPayloadBuild.py#L254

While we are targeting the UPL support for EDK2, I'd prefer if it also works for other payloads. IMO that means following the spec.

Reply:
A new patch will be provided to support this Position Independent Execution feature, and this patch has been validated at my local, UPL FIT image can be loaded into any aligned available address. Further more, with this new patch, u-root works with top of tree of EDK2 code base without any changes.
We still need grant from your side to merge this PR first, since upcoming patches are all based on this PR.